feat: add worker library

joshuarubin opened the pull request:
Add the worker library according to the challenge requirements and design doc.
Additionally, include the generated stubs from the proto files.

~~Note that one of the major tradeoffs I've made is that each instance of the `io.ReadCloser` returned by `JobOutput()` keeps a full copy of the job output and continues to append all new incoming data to it. Once data has been read by clients of the reader, it does not need to be retained any longer. Therefore, as presently implemented, the readers are quite inefficient and can use quite a lot of memory, especially if many readers are created for a job.~~

tigrato commented on internal/config/config.go:
```diff
@@ -0,0 +1,18 @@
+
+import "time"
+
+type TLS struct {
```
Missing godocs

tigrato commented on internal/config/config.go:
```diff
@@ -0,0 +1,18 @@
+	KeyFile    string
+}
+
+type Config struct {
```
missing godocs

tigrato commented on internal/config/config.go:
```diff
@@ -0,0 +1,18 @@
+	Addr            string
+	TLS             TLS
+	ShutdownTimeout time.Duration
+	CPUMax          string
+	MemoryMax       string
+	IOMax           []string
```
Should these properties be a string?

joshuarubin replied:
The challenge requirements said not to spend much time on config, so I chose the simplest thing possible here which was to pass in the literal values that would be used in the cgroup files. Probably not what I would choose in any other case, but it is easy and works.

joshuarubin replied:
Updated with better typing and validation

tigrato commented on internal/job/id.go:
```diff
@@ -0,0 +1,15 @@
+
+type Prefix struct{}
+
+func (Prefix) Prefix() string { return "job" }
```
Is it required to add a prefix? Aren't generally available UUIDs good enough?


also add comments/godoc

joshuarubin replied:
I've found it really helpful when dealing with large systems, that may have lots of different kinds of identifiers, to ensure that they are prefixed. When digging through logs, it may not be clear what a UUID is, but prefixed IDs stand out. Also, these IDs are K-sortable and based on the upcoming UUID-7 standard, so they're not _that_ different.

tigrato commented on internal/job/job.go:
```diff
@@ -0,0 +1,223 @@
+
+	j.cmd.SysProcAttr = sysProcAttr()
+
+	j.cmd.Env = os.Environ()
+	j.cmd.Env = append(j.cmd.Env, env...)
```
```suggestion
	j.cmd.Env = append(os.Environ(), env...)
```

tigrato commented on internal/job/job.go:
```diff
@@ -0,0 +1,223 @@
+		id:     id,
+		userID: userID,
+		done:   done,
+		cmd:    exec.CommandContext(ctx, command, args...),
```
Is it a good design to have the job depending on the parent context? If the parent context is the RPC context, won't it be canceled as soon as the job start RPC ends?

tigrato commented on internal/job/job.go:
```diff
@@ -0,0 +1,223 @@
+	return &j, nil
+}
+
+var ErrAlreadyStarted = errors.New("already started")
```
godoc

tigrato commented on internal/job/job.go:
```diff
@@ -0,0 +1,223 @@
+// setStatus sets the job status. status can only move to higher values:
+// not_started -> running -> complete -> stopped
+func (j *Job) setStatus(st Status) {
+	for {
```
Is this for loop necessary? Looking at the implementation and operation ordering, I wonder if store alone solves all the issues?

joshuarubin replied:
The concern is if the job is manually stopped at the same moment that it completes. If the job goes first from `StatusRunning` then to `StatusCompleted` in between the time that the `setStatus(StatusStopped)` gets the current status (as `StatusRunning`) and then does the `CompareAndSwap`, which fails because it is now `StatusCompleted`. The state should rightfully be changed to `StatusStopped` as the "higher" value, but without the loop, it won't be.

joshuarubin replied:
After further review, I agree with both your assessment. StatusCompleted is set just before the `done` channel is closed and `Stop()` waits until after the `done` channel is closed to set the status. As such, there is no possibility of a race. I'll update. Thanks for the sharp eyes.

joshuarubin replied:
And with that, status doesn't even need to be atomic either.

tigrato commented on internal/job/job.go:
```diff
@@ -0,0 +1,223 @@
+	startOnce sync.Once
+
+	stopOnce sync.Once
+	stopErr  atomic.Value
```
does this need to be atomic?

joshuarubin replied:
If `Stop()` is called concurrently, one will "win" and execute the `Once` func and the other will read the value of `stopErr` while the `Once` func may be writing it. So yes, it does need coordination, but there is still a bug. In the case I described, the "loser" that doesn't run the `Once` func may erroneously return a nil error because it read the value before the `Once` func hasn't yet written it. It probably needs a `RWMutex` instead.

joshuarubin replied:
Scratch what I just said. I forgot how Once works. Concurrent calls to `Once.Do()` all block until the "one" func running returns, so there isn't a concurrency issue here. Will make it a var.

tigrato commented on internal/user/user.go:
```diff
@@ -0,0 +1,7 @@
+package user
+
+type ID string
```
godoc

tigrato commented on internal/safebuffer/safebuffer.go:
```diff
@@ -0,0 +1,88 @@
+		default:
+			v := make([]byte, len(p))
+			copy(v, p)
+			dataCh <- v
```
What will happen if a reader is slow to consume the output?

joshuarubin replied:
The reader's `receiver` runs in its own goroutine where it writes to its own buffer. It does need to take a lock when writing so there is no contention while the reader's buffer is being read by clients. It holds the lock for the shortest period possible while reading and writing to the reader buffer. While there may be a very brief stutter, it is never be dependent on external/uncontrollable factors (like the client reading the `io.ReadCloser`) as to how quickly it can be consumed.

tigrato commented on internal/safebuffer/safereader/safereader.go:
```diff
@@ -0,0 +1,160 @@
+	r := Reader{
+		close:  closeFn,
+		closed: closed,
+		buf:    bytes.NewBuffer(buf),
```
Why do we need to keep a full copy of the job output per reader?

If a job outputs 1Gb of data and we have 5 active readers, we will take 5GB of ram. Is there a way where we do not need to implement a solution with such bottleneck ?

codingllama replied:
+1

joshuarubin replied:
I tried to call this out in the pull request's description, but I should also explain it in the code. This is by far the biggest compromise I've made in the design. Keeping a full copy of the output buffer from the time that a reader is created is a very naive approach that is very memory and time inefficient. I did this to keep the code simple and with the understanding the the challenge was given with "unlimited cpu and unlimited memory". This is not a design I would ever propose for production.

I think what I would have liked to do instead is use a `RWMutex` on the original buffer and let the readers read directly from that and maintain their own position/index. The problem I had when thinking through this is that if I imagined a lot of connected readers holding `RLock`s, then writes may block for unacceptably long times. I may be able to fix this by limiting the time that `RLock`s are held and looping, thereby giving an opportunity for the write locks to be acquired. I worried that trying to figure out how to do this correctly and without races would become quite complex. In keeping with the spirit of the challenge, "do the simplest thing possible that satisfies the requirements", I chose what I thought to be the simpler option to implement.

There is another optimization I could also do to the current implementation, which would be to discard data from readers once it has been read. Again, it's more work and not necessarily a requirement, so I did not choose to implement this either.

tigrato commented on internal/safebuffer/safereader/safereader.go:
```diff
@@ -0,0 +1,160 @@
+		r.dataCh = dataCh
+		r.dataAvail = make(chan struct{})
+
+		go r.receiver()
```
is it possible to live without spinning extra go-routines?

joshuarubin replied:
With the current design, it's one goroutine per reader, so it's not awful, but it is required to prevent blocking/slow writes. I suppose I could have used buffered channels to fix _most_ blocking issues, but it's still not a guarantee that they will always be avoided. Generally, I view buffered channels as a bit of a smell (except in some cases where the buffer is 1 so that writes don't block forever if a reader disappears) that the code could probably be improved. Besides, the cost of launching a single goroutine is very small, it needs to be managed properly, but I don't see it as a major concern.

tigrato commented on internal/safebuffer/safereader/safereader.go:
```diff
@@ -0,0 +1,160 @@
+
+		go r.receiver()
+	} else {
+		r.Close()
```
why closing the reader here?

joshuarubin replied:
Readers are closed when jobs complete (or when clients call `Close()`), but in this case the job has already completed and we know no new job output will be coming in. As a result, the state of the reader is simplified in that it only needs to be a buffer for the data it has when it is created. It doesn't need to be added to the map in the `SafeBuffer`, it doesn't need a goroutine, the calls to `Read()` can just defer to the internal buffer. By closing it immediately, ensure that clients won't block indefinitely for jobs that have already completed and for which they haven't called `Close()` in a separate goroutine. I'll add a comment to clarify.

tigrato commented on internal/safebuffer/safereader/safereader.go:
```diff
@@ -0,0 +1,160 @@
+	n, err := r.buf.Write(p)
+
+	if err == nil && n > 0 {
+		close(r.dataAvail)
```
if `dataAvail` is only consumed in a single place, why do you need to close (broadcast) info everywhere?

joshuarubin replied:
I'm not quite sure what you mean by "everywhere". `write()` and `Read()` could be called from different goroutines. While this method can be used to signal between many goroutines, it's no less effective when signaling to just a single other one. It seems like you have something else in mind though and I'm curious what it may be...

joshuarubin replied:
I see. However, then you've got an issue if you have more pending sends than the buffer size. This pattern won't ever block the sender.

tigrato commented on internal/safebuffer/safereader/safereader.go:
```diff
@@ -0,0 +1,160 @@
+		// the reader was closed, so we return io.EOF
+		return 0, io.EOF
+	case <-dataAvail:
+		// more data was written to buf, so let's return that
+		r.mu.Lock()
+		defer r.mu.Unlock()
+		return r.buf.Read(p)
```
It will also return io.EOF and mislead the client because we ended up the buffer but the job hasn't ended

joshuarubin replied:
yeah, definitely a bug here, nice catch!

codingllama commented on internal/worker/jobworker/jobworker.go:
```diff
@@ -0,0 +1,186 @@
+	jobs map[job.ID]*job.Job
+}
+
+// ensure JobWorker implements the Worker interface
+var _ worker.Worker = (*JobWorker)(nil)
```
Could we ditch the interface and expose JobWorker directly?

The interface is useful to express the API in the design phase, but I don't think the API actually calls for it.

- https://go.dev/wiki/CodeReviewComments#interfaces

joshuarubin replied:
Fine with me!

codingllama commented on internal/worker/jobworker/jobworker.go:
```diff
@@ -0,0 +1,186 @@
+type ReexecCommand struct {
+	Command string   // often "/proc/self/exe"
+	Args    []string // usually a command that puts this binary in a different "mode", e.g. "child" or "helper"
+	Env     []string // additional env variables, in the form of "key=value", to be added to the current os.Environ()
```
Ditch Env?

joshuarubin replied:
It was required for testing unfortunately

codingllama commented on internal/worker/jobworker/jobworker.go:
```diff
@@ -0,0 +1,186 @@
+
+type cgroup struct {
+	cfg        *config.Config
+	createOnce sync.Once //nolint:unused,nolintlint
+	createErr  error     //nolint:unused,nolintlint
+	name       string    //nolint:unused,nolintlint
```
nit: Write a proper nolint comment. There's not much point in using nolintlint if we ignore it.

joshuarubin replied:
This was a PITA. It is not a lint error on linux, but is a lint error on mac. This was the only way I could find to quiet it in both cases.

joshuarubin replied:
Fixed by minimizing the amount in the `_linux.go` file so the linter would see that these fields are used.

codingllama commented on internal/worker/jobworker/jobworker.go:
```diff
@@ -0,0 +1,186 @@
+var _ worker.Worker = (*JobWorker)(nil)
+
+// New creates a new JobWorker
+func New(cfg *config.Config, reexec *ReexecCommand) *JobWorker {
```
Same remarks about validating the params and letting them be mutated from the outside.

joshuarubin replied:
Added validation. Changed config fields (and put it in worker package). Copy it so it isn't externally mutable.

codingllama commented on internal/worker/jobworker/jobworker.go:
```diff
@@ -0,0 +1,186 @@
+	cmd, err := exec.LookPath(command)
+	if err != nil {
+		err = fmt.Errorf("LookPath error: %w", err)
+		fmt.Fprintln(os.Stderr, err)
```
Use a proper logger, like slog. Same for other printfs.

codingllama commented on internal/worker/jobworker/jobworker.go:
```diff
@@ -0,0 +1,186 @@
+func (w *JobWorker) StartJobChild(_ context.Context, command string, args ...string) error {
+	cmd, err := exec.LookPath(command)
+	if err != nil {
+		err = fmt.Errorf("LookPath error: %w", err)
```
nit: The uppercase error will trip various linters.

- https://go.dev/wiki/CodeReviewComments#error-strings

codingllama commented on internal/worker/jobworker/jobworker.go:
```diff
@@ -0,0 +1,186 @@
+}
+
+// JobStatus will return the JobStatus and optional ExitCode from the job. The
+// ExitCode will not exist if the job is still running, or if it exited
+// uncleanly, e.g. because it was killed by a signal. If the job does not exist,
```
> // ExitCode will not exist (...) if it exited uncleanly, e.g. because it was killed by a signal.

Is this correct?

joshuarubin replied:
I thought it was when I wrote it, then my testing showed that it wasn't. I'll update.

codingllama commented on internal/worker/jobworker/jobworker.go:
```diff
@@ -0,0 +1,186 @@
+// circumstance and should never be called in any other situation. It will
+// create a new cgroup with cpu, memory and io limits applied, then it will
+// remount /proc and finally it will execute the command, with optional args.
+func (w *JobWorker) StartJobChild(_ context.Context, command string, args ...string) error {
```
Drop the context.Context from the signature? It seems to serve no purpose.

Same for others.

joshuarubin replied:
You never know when you'll need a context later... but I hear you. Will update.

codingllama replied:
True. We have the luxury of having a fixed scope here.

codingllama commented on internal/worker/jobworker/jobworker_linux.go:
```diff
@@ -0,0 +1,91 @@
+// the values for cpu.max, memory.max and io.max. it also adds the current
+// process's pid to cgroup.procs.
+func (g *cgroup) createLeaf() error {
+	g.createOnce.Do(func() {
```
Is the sync.Once necessary here?

joshuarubin replied:
Yes, I don't want to keep creating container/top-level cgroups each time I make a leaf cgroup. This ensures I ever only make one root cgroup per worker.

codingllama commented on internal/worker/jobworker/jobworker_linux.go:
```diff
@@ -0,0 +1,91 @@
+// process's pid to cgroup.procs.
+func (g *cgroup) createLeaf() error {
+	g.createOnce.Do(func() {
+		g.createErr = g.createRoot()
```
Why a struct field for createErr? Why not a `var`?

joshuarubin replied:
because if creating the root cgroup failed, and because it's in the once func we won't try again, we need to abort further attempts to make leaf cgroups.

joshuarubin replied:
`syscall.Mount()` isn't available on all GOOS, so I'll still need to use build flags to protect that, but I can move the rest out.

codingllama commented on internal/worker/jobworker/jobworker_linux.go:
Suggestion: This file is all file-manipulation, so while the logic is linux-specific you could drop the build constraint, save yourself a couple of `//nonlints` and even test it regardless of platform.

joshuarubin replied:
yeah, this is a great suggestion, let me see what I can do...

joshuarubin replied:
`syscall.Mount()` isn't available on all GOOS, so I'll still need to use build flags to protect that, but I can move the rest out.

codingllama commented on internal/worker/jobworker/jobworker_linux.go:
```diff
@@ -0,0 +1,91 @@
+
+	for _, v := range cgroupData {
+		if err = os.WriteFile(filepath.Join(subCgroup, v.file), []byte(v.value), filePerm); err != nil {
+			return err
```
Is it worth adding context to this error, like the file we are writing?

joshuarubin replied:
sounds good to me

codingllama commented on internal/worker/jobworker/jobworker_linux.go:
```diff
@@ -0,0 +1,91 @@
+
+func (w *JobWorker) prepJobEnv() error {
+	if err := w.cgroup.createLeaf(); err != nil {
+		return err
```
nit: Add context to the error?

codingllama commented on internal/worker/jobworker/jobworker_linux.go:
```diff
@@ -0,0 +1,91 @@
+
+	cg, err := os.MkdirTemp(prefix, "job-worker-")
+	if err != nil {
+		return err
```
nit: Add context to the error?

codingllama commented on internal/worker/jobworker/jobworker_other.go:
```diff
@@ -0,0 +1,8 @@
+//go:build !linux
+// +build !linux
```
I think you can drop these considering we require Go 1.22.

```suggestion
```

joshuarubin replied:
This was something `gopls` did automatically. ¯\_(ツ)_/¯

codingllama replied:
I thought tooling was past that already, my bad.

codingllama commented on internal/safebuffer/safereader/safereader.go:
```diff
@@ -0,0 +1,160 @@
+	r.mu.Lock()
+
+	n, err := r.buf.Read(p)
+	if err != io.EOF || r.isClosed() {
```
To be safe:

```suggestion
	if errors.Is(err, io.EOF) || r.isClosed() {
```

joshuarubin replied:
good catch

codingllama commented on internal/safebuffer/safereader/safereader.go:
```diff
@@ -0,0 +1,160 @@
+// called.
+func (r *Reader) Close() error {
+	r.closeOnce.Do(r.close)
+	return nil
```
Should we error on multiple closes?

Document the chosen behavior?

joshuarubin replied:
multiple closes are actually expected. the job completing will cause a close, and clients can choose to close early too.

codingllama commented on internal/safebuffer/safereader/safereader.go:
```diff
@@ -0,0 +1,160 @@
+				return
+			}
+			if _, err := r.write(data); err != nil {
+				slog.Error("error writing to Reader buffer", "err", err)
```
nit:

```suggestion
				slog.Error("Error writing to Reader buffer", "err", err)
```

codingllama commented on internal/safebuffer/safereader/safereader.go:
```diff
@@ -0,0 +1,160 @@
+// Reader is a goroutine safe io.ReadCloser that makes initial data available to
+// callers as well as new data that arrives later via channel.
+type Reader struct {
+	mu        sync.Mutex
+	buf       *bytes.Buffer
+	close     func()
+	closed    <-chan struct{}
+	dataCh    <-chan []byte
+	dataAvail chan struct{}
+	closeOnce sync.Once
```
There are a lot of concurrency control structures here too, I wonder if we could simplify.

codingllama commented on internal/safebuffer/safereader/safereader.go:
```diff
@@ -0,0 +1,160 @@
+	copy(buf, data)
+
+	r := Reader{
+		close:  closeFn,
```
Why is this a field on Reader?

joshuarubin replied:
The func is "how" to close and signal to release resources, the chan is how the reader can know that it has been signaled to close to actually break out of blocking conditions and exit goroutines.

codingllama commented on internal/safebuffer/safebuffer.go:
```diff
@@ -0,0 +1,88 @@
+	mu      sync.RWMutex
+	buf     bytes.Buffer
+	jobDone <-chan struct{}
+	chans   map[chan []byte]<-chan struct{}
```
Should this be a slice of Channels instead?

joshuarubin replied:
The buffer needs to know if a connected reader has been closed so it doesn't attempt to send to a closed channel causing a panic. Are you suggesting using something like this instead?:
```go
[]struct{
    DataCh chan []byte
    DoneCh ←chan struct{}
}
```

codingllama replied:
I meant `[]*safereader.Channels`.

codingllama commented on internal/safebuffer/safebuffer_test.go:
```diff
@@ -0,0 +1,203 @@
+		for range 3 {
+			n, err = r0.Read(b)
+			require.NoError(err)
+			assert.Equal(3, n)
```
Assert the content read?

codingllama commented on internal/safebuffer/safebuffer_test.go:
```diff
@@ -0,0 +1,203 @@
+			t.Fatal("expected Read not to complete because jobDone wasn't closed")
+		}
+
+		close(jobDone)
+
+		timer.Reset(10 * time.Millisecond)
+		select {
+		case <-timer.C:
+			t.Fatal("expected Read to complete because jobDone was closed")
+		case <-done:
+		}
+		assert.Equal(io.EOF, err)
+		assert.Equal(0, n)
```
What is this block testing? `done` is already closed and `n` and `err` are leftovers, or am I missing something?

joshuarubin replied:
It's checking to make sure that the reader blocks until `io.EOF` is received. Here's the sequence:
1. Write to the buffer
2. Synchronous read everything from the buffer
3. Try reading more from the buffer in the goroutine
4. This times out because nothing more is written
5. Close the `jobDone` channel to simulate the job completing
6. Now the reader should should return `io.EOF` and the `done` channel will close

I'll clarify in the code

joshuarubin commented on internal/job/job.go:
```diff
@@ -0,0 +1,223 @@
+	cmd    *exec.Cmd
+	buf    *safebuffer.Buffer
+
+	startOnce sync.Once
+
+	stopOnce sync.Once
+	stopErr  atomic.Value
+
+	status atomic.Uint32
```
Yeah, I think you may be right, I'll review

joshuarubin replied:
`stopErr` doesn't need to be atomic and has been updated. however, the `Once` funcs aren't really something mutexes alone can replicate. that leaves the atomic `status` which i could use a mutex on a standard status value, but that wouldn't do all that much to clean things up and atomics are certainly faster than mutex locks.

joshuarubin replied:
I was able to remove the need for the atomic on the status field as well.

joshuarubin commented on internal/safebuffer/safebuffer.go:
```diff
@@ -0,0 +1,88 @@
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	for dataCh := range b.chans {
+		close(dataCh)
+		delete(b.chans, dataCh)
+	}
```
I thought a lot about this concern, and tried to test it too. Based on what I understand from the docs and the results of my test, no, but I'm still not fully convinced either.

The [docs](https://pkg.go.dev/os/exec#Cmd.Wait) for `cmd.Wait()` say:
> Wait waits for the command to exit and waits for any copying to stdin or copying from stdout or stderr to complete.

Let's presume we're not at these lines because the `ctx` became done (in which case all bets are off). So in order to get to the loop you highlighted, the `jobDone` channel must have been closed. That channel, necessarily, can only be closed after `Wait()` completes. This means that all `Write()` calls to the `SafeBuffer`, which are synchronous, must complete before `dataCh` can be closed. The `SafeReader` may receive data via channel in a separate goroutine, which, once received unblocks the `Write()` call in the `SafeBuffer`. `Write()` holds the mutex for the entire time it executes, and so does `jobWatcher()` when it gets to that loop. This means that all of the channels will receive all of their data before `close()` is called on them and so all of the data will make it through the channel.

joshuarubin commented on internal/job/job.go:
```diff
@@ -0,0 +1,223 @@
+	userID user.ID,
+	command string,
+	args []string,
+	env []string,
```
I didn't have it initially, but added it in order to make it possible to test.

joshuarubin commented on internal/job/job.go:
```diff
@@ -0,0 +1,223 @@
+	var started bool
+	j.startOnce.Do(func() {
+		j.setStatus(StatusRunning)
+		err = j.cmd.Start()
```
Yeah, this is bad error handling. Will correct.

joshuarubin commented on internal/job/job.go:
```diff
@@ -0,0 +1,223 @@
+				j.setStatus(StatusComplete)
+				close(j.done)
+			}()
+			j.cmdErr = j.cmd.Wait()
```
Sounds like a good idea

joshuarubin commented on internal/config/config.go:
```diff
@@ -0,0 +1,18 @@
+type TLS struct {
+	CACertFile string
+	CertFile   string
+	KeyFile    string
```
We'll deal with these in an upcoming PR when they're required. Removed from this PR.

joshuarubin commented on internal/config/config.go:
```diff
@@ -0,0 +1,18 @@
+}
+
+type Config struct {
+	Addr            string
```
We'll deal with this in an upcoming PR when it's required. Removed from this PR.

joshuarubin commented on internal/job/job.go:
```diff
@@ -0,0 +1,223 @@
+
+// Job represents a system process
+type Job struct {
+	cfg    *config.Config
```
1. Yes, I've made a config specifically for the worker now
2. I'm now cloning the config to make it externally immutable

joshuarubin commented on internal/config/config.go:
```diff
@@ -0,0 +1,18 @@
+	ShutdownTimeout time.Duration
+	CPUMax          string
+	MemoryMax       string
+	IOMax           []string
```
Updated with:
``` go
CPUMax        float32  // the maximum cpu usage as a decimal, 0 < value <= 1, 0 indicates no max
MemoryMax     uint32   // the maximum memory usage in bytes, 0 indicates no max
RIOPSMax      uint32   // the maximum read io operations per second, 0 indicates no max
WIOPSMax      uint32   // the maximum write io operations per second, 0 indicates no max
```

joshuarubin commented on internal/job/job.go:
```diff
@@ -0,0 +1,223 @@
+func New(
+	ctx context.Context,
+	cfg *config.Config,
+	userID user.ID,
```
Added a check to ensure user id isn't empty. Config is handled differently now too.

codingllama commented on pkg/worker/worker.go:
```diff
@@ -0,0 +1,382 @@
+	scanner := bufio.NewScanner(f)
+	for scanner.Scan() {
+		line := scanner.Text()
+		// the lines we're looking for look like:
+		//  253        0  104857600 vda
```
Nice, I frequently ask for this comment.

joshuarubin commented on pkg/job/job.go:
```diff
@@ -0,0 +1,230 @@
+func (j *Job) wait() {
+	defer func() {
+		j.setStatus(StatusCompleted)
+		close(j.done)
```
I have a note in the job struct definition and I'm adding one here too

joshuarubin commented:
* I've updated the output buffer so that it no longer copies data to each reader
* The output buffer is now implemented with an immutable linked list
* The output buffer does need a RWMutex, but it is held very briefly and never when walking the linked list
